src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/StockInMemoryRepositoryImpl.kt

```diff
@@ -13,40 +14,84 @@ import java.util.*
 class StockInMemoryRepositoryImpl : StockRepository {
     private val db: HashMap<UUID, Stock> = HashMap()
 
-    override fun insertAll(stocks: List<Stock>): MutableCollection<Stock> {
+    override fun insertAll(stocks: List<Stock>): List<Stock> {
         stocks.forEach {
             val id = UUID.randomUUID()
             it.id = id
             db[id] = it
         }
-        return stocks.toMutableList()
+        return stocks
     }
 
-    override fun findFiveBestStocksToBuy(): List<Pair<String?, List<Stock>>> {
+    override fun findTopNStocks(n: Int): List<Pair<String?, List<Stock>>> {
         if (db.isEmpty()) return emptyList()
 
-        val maxPercentChange = db.asSequence().mapNotNull { it.value.percentChange }.maxBy { it }
-        val maxChange = db.asSequence().mapNotNull { it.value.change }.maxBy { it }
-
-        return db.asSequence().filterNot { it.value.symbol == null }.filter { stock ->
-            stock.value.dateOfRetrieval?.isBefore(Instant.now().plus(Duration.ofHours(1))) == true
-        }.map { it.value }.groupBy { it.symbol }.toList().sortedByDescending { (_, stockList) ->
-            val avgChange = stockList.mapNotNull { it.change }
-                .fold(BigDecimal.ZERO) { acc, bigDecimal -> acc + bigDecimal } / BigDecimal.valueOf(stockList.size.toLong())
-            val avgPercentChange = stockList.mapNotNull { it.percentChange }
-                .fold(BigDecimal.ZERO) { acc, bigDecimal -> acc + bigDecimal } / BigDecimal.valueOf(stockList.size.toLong())
-
-            ((avgChange / maxChange) * BigDecimal.valueOf(0.5)) + ((avgPercentChange / maxPercentChange) * BigDecimal.valueOf(
-                0.5
-            ))
-        }.take(5)
+        val maxPercentChange = getMaxBigDecimal { stock: Stock -> stock.percentChange }
+        val maxChange = getMaxBigDecimal { stock: Stock -> stock.change }
+
+        val timeOfRequest = Instant.now()
+
+        return db.asSequence()
+            .filterNot { (_, stock) -> stock.symbol == null }
+            .filter { (_, stock) ->
+                isStockInValidDateRange(stock, timeOfRequest)
+            }
+            .map { it.value }
+            .groupBy { it.symbol }
+            .toList()
+            .sortedByDescending(compareByPrice(maxChange, maxPercentChange))
+            .take(n)
             .map { (symbol, stocks) ->
                 symbol to stocks.distinctBy { it.currentPrice }
-                    .sortedByDescending { stock -> stock.dateOfRetrieval }.take(5)
+                    .sortedByDescending { stock -> stock.dateOfRetrieval }
```

Comment:
```suggestion
                    .sortedByDescending { it.dateOfRetrieval }
```
here IMO you don't need explicit variable and 'it' is enough